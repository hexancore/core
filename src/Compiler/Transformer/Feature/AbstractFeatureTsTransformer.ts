import ts from 'typescript';
import type { FeatureMap, FeatureMeta } from "../../../Util/Feature/Meta/FeatureMeta";
import { ModuleClassTsTransformer } from '../ModuleClassTsTransformer';
import type { TsImportHelper } from '../TsImportHelper';
import type { FeatureTransformContext } from './FeatureTransformContext';
import type { FeatureSourcePath } from "../../../Util/Feature/FeatureModuleDiscoverer";

export interface FeatureTsTransformerHelpers {
  importHelper: TsImportHelper;
  moduleClassTransformer: ModuleClassTsTransformer;
  features: FeatureMap;
}

export abstract class AbstractFeatureTsTransformer {
  protected importHelper: TsImportHelper;
  protected moduleClassTransformer: ModuleClassTsTransformer;
  protected features: FeatureMap;

  public constructor(helpers: FeatureTsTransformerHelpers) {
    this.importHelper = helpers.importHelper;
    this.moduleClassTransformer = helpers.moduleClassTransformer;
    this.features = helpers.features;
  }

  public abstract transform(source: ts.SourceFile, context: FeatureTransformContext): ts.SourceFile;

  public abstract supports(featureSourcePath: FeatureSourcePath, feature: FeatureMeta): boolean;

}