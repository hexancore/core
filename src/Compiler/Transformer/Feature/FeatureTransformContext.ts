import type ts from "typescript";
import type { FeatureMeta } from "../../../Util/Feature/Meta/FeatureMeta";

export interface FeatureTransformContext {
  feature: FeatureMeta;
  sourceFilePathWithoutRoot: string;
  tsContext: ts.TransformationContext;
  diagnostics: ts.Diagnostic[];
}