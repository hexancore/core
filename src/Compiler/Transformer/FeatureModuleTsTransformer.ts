import { execFileSync, execSync } from "node:child_process";
import ts from 'typescript';
import { type FeatureApplicationMessageInfo, type FeatureModuleDiscovery } from "../../Util/FeatureModuleDiscoverer";
import { TsTransfromerHelper } from "./TsTransformerHelper";

/**
 * Adding automatic injection of message handlers, services, infra module to `[Feature]Module` source.
 * Less write, more fun !
 */
export class FeatureModuleTsTransformer {
  public constructor(private features: Map<string, FeatureModuleDiscovery>) {
  }

  public static create(sourceRoot: string, compilerDir: string): FeatureModuleTsTransformer {
    const features = this.loadFeaturesMapSync(sourceRoot, compilerDir);
    return new this(features);
  }

  private static loadFeaturesMapSync(sourceRoot: string, compilerDir: string): Map<string, FeatureModuleDiscovery> {
    // TS transformer can't use async, but we need async code and execSync allows run async code :) Hack of day !
    const output = execFileSync('node', [`${compilerDir}/discoverer.js`, sourceRoot]);
    const features = JSON.parse(output.toString());
    return new Map(features);
  }

  public transform(featureName: string, source: ts.SourceFile, context: ts.TransformationContext): ts.SourceFile {
    const feature = this.features.get(featureName);
    if (!feature) {
      throw new Error(`No found feature module info, name: ${featureName}`);
    }

    let imported = false;
    const currentImports = new Set<string>();
    let importMap: Map<string, ts.ImportDeclaration>;

    const visitor: ts.Visitor = (node) => {
      if (ts.isImportDeclaration(node)) {
        currentImports.add(TsTransfromerHelper.extractModuleSpecifierFromImportDeclaration(node));
        return node;
      }

      if (!ts.isImportDeclaration(node) && !imported) {
        imported = true;
        importMap = this.createImportDeclarations(feature, currentImports);
        if (ts.isClassDeclaration(node)) {
          node = ts.visitEachChild(node, (n) => this.transformModuleDecorator(n, feature, importMap, context), context);
        }
        return [...importMap.values(), node];
      }

      if (ts.isClassDeclaration(node)) {
        node = ts.visitEachChild(node, (n) => this.transformModuleDecorator(n, feature, importMap, context), context);
      }

      return node;

    };

    return ts.factory.updateSourceFile(source, ts.visitNodes(
      source.statements,
      visitor,
      (node): node is ts.Statement => ts.isStatement(node))
    );
  }

  private createImportDeclarations(feature: FeatureModuleDiscovery, currentImports: Set<string>) {
    const importMap: Map<string, ts.ImportDeclaration> = new Map();

    this.addMessageHandlerImportDeclarations(feature.application.commands, currentImports, importMap);
    this.addMessageHandlerImportDeclarations(feature.application.queries, currentImports, importMap);
    this.addMessageHandlerImportDeclarations(feature.application.events, currentImports, importMap);
    this.addServiceImportDeclarations(feature, currentImports, importMap);
    this.addInfrastructureModuleImportDeclaration(feature, importMap);

    return importMap;
  }

  private addMessageHandlerImportDeclarations(messages: FeatureApplicationMessageInfo[], currentImports: Set<string>, importMap: Map<string, ts.ImportDeclaration>) {
    for (const m of messages) {
      const importPath = `./${m.path}/${m.handlerClassName}`;
      if (currentImports.has(importPath)) {
        continue;
      }

      importMap.set(m.handlerClassName, this.createImportDeclaration(m.handlerClassName, importPath));
    }
  }

  private addServiceImportDeclarations(feature: FeatureModuleDiscovery, currentImports: Set<string>, importMap: Map<string, ts.ImportDeclaration>): void {
    for (const s of feature.application.services) {
      if (s.isInjectable) {
        const importPath = `./${s.path}`;
        if (currentImports.has(importPath)) {
          continue;
        }

        importMap.set(s.className, this.createImportDeclaration(s.className, importPath));
      }
    }
  }
  private addInfrastructureModuleImportDeclaration(feature: FeatureModuleDiscovery, importMap: Map<string, ts.ImportDeclaration>): void {
    const infrastructureModuleImportDec = this.createImportDeclaration(feature.infrastructure.module.className, `./${feature.infrastructure.module.path}`);
    importMap.set(feature.infrastructure.module.className, infrastructureModuleImportDec);
  }

  private createImportDeclaration(importName: string, importPath: string): ts.ImportDeclaration {
    const namedImports = ts.factory.createNamedImports([
      ts.factory.createImportSpecifier(false, undefined, ts.factory.createIdentifier(importName))
    ]);
    const importFrom = ts.factory.createStringLiteral(importPath);
    return ts.factory.createImportDeclaration(
      undefined,
      ts.factory.createImportClause(false, undefined, namedImports),
      importFrom
    );
  }

  private transformModuleDecorator(
    node: ts.Node,
    feature: FeatureModuleDiscovery,
    importMap: Map<string, ts.ImportDeclaration>,
    context: ts.TransformationContext
  ) {
    return TsTransfromerHelper.transformModuleDecoratorExpression(node, () => {
      const imports = [
        feature.infrastructure.module.className
      ];

      const providers = [
        ...feature.application.commands.map((m) => m.handlerClassName),
        ...feature.application.queries.map((m) => m.handlerClassName),
        ...feature.application.events.map((m) => m.handlerClassName),
        ...feature.application.services.filter((s) => s.isInjectable).map((s) => s.className)
      ];
      const classIdentifierMapper = (className) => TsTransfromerHelper.createNamedImportAccess(importMap.get(className)!, className, context);
      return {
        imports: imports.map(classIdentifierMapper),
        providers: providers.map(classIdentifierMapper),
      };
    });
  }
}