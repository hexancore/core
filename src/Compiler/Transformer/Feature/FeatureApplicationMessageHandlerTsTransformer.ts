import { FeatureApplicationCommandMeta, type FeatureApplicationQueryMeta } from '../../../Util/Feature/Meta';
import ts from "typescript";
import { type FeatureMeta } from "../../../Util/Feature/Meta";
import { AbstractFeatureTsTransformer } from "./AbstractFeatureTsTransformer";
import type { FeatureTransformContext } from "./FeatureTransformContext";
import type { FeatureSourcePath } from "../../../Util/Feature/FeatureModuleDiscoverer";
import { ImportDeclarationWrapper } from "../Helper/ImportDeclarationWrapper";

interface VisitContext extends FeatureTransformContext {
  meta: FeatureApplicationCommandMeta | FeatureApplicationQueryMeta;
  handlerDecoratorName: string;
  cqrsImport: ImportDeclarationWrapper;
}

export class FeatureApplicationMessageHandlerTsTransformer extends AbstractFeatureTsTransformer {

  public supports(featureSourcePath: FeatureSourcePath, feature: FeatureMeta): boolean {
    const meta = feature.applicationCqrsHandlerMap.get(featureSourcePath.localSourcePath);
    return !!meta;
  }

  public transform(source: ts.SourceFile, context: FeatureTransformContext): ts.SourceFile {
    const messageMeta = context.feature.applicationCqrsHandlerMap.get(context.featureSourcePath.localSourcePath)!;
    const handlerDecoratorName = messageMeta instanceof FeatureApplicationCommandMeta ? "CommandHandler" : "QueryHandler";
    const cqrsImport = ImportDeclarationWrapper.create(handlerDecoratorName, "@nestjs/cqrs", context.tsContext);
    const visitContext: VisitContext = {
      ...context,
      meta: messageMeta,
      handlerDecoratorName: handlerDecoratorName,
      cqrsImport
    };

    const transformed = ts.factory.updateSourceFile(source, [
      cqrsImport.declaration,
      ...ts.visitNodes(
        source.statements,
        this.visitSourceStatement.bind(this, visitContext),
        (node): node is ts.Statement => ts.isStatement(node)
      )
    ]);

    return transformed;
  }

  private visitSourceStatement(visitContext: VisitContext, node: ts.Statement) {
    if (ts.isClassDeclaration(node)) {
      const handlerDecorator = ts.factory.createDecorator(ts.factory.createCallExpression(
        visitContext.cqrsImport.get(visitContext.handlerDecoratorName),
        undefined,
        [ts.factory.createIdentifier(visitContext.meta.className)]
      ));
      return ts.factory.updateClassDeclaration(
        node,
        [handlerDecorator, ...node.modifiers ?? []],
        node.name,
        node.typeParameters,
        node.heritageClauses,
        node.members
      );
    }

    return node;
  }
}