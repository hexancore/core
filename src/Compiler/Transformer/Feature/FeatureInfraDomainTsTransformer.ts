import ts from 'typescript';
import type { FeatureMeta } from "../../../Util/Feature/Meta/FeatureMeta";
import { AbstractFeatureTsTransformer } from './AbstractFeatureTsTransformer';
import { TsTransfromerHelper } from '../TsTransformerHelper';
import type { AddImportTransformDef } from '../ModuleClassTsTransformer';
import type { FeatureTransformContext } from './FeatureTransformContext';

export class FeatureInfraDomainModuleTsTransformer extends AbstractFeatureTsTransformer {

  public transform(source: ts.SourceFile, context: FeatureTransformContext): ts.SourceFile {
    const feature = context.feature;
    const domainErrorsClassName = feature.name + "DomainErrors";

    const imports: AddImportTransformDef[] = [
      { name: "DomainInfraModuleHelper", importModuleSpecifier: '@hexancore/core' },
      { name: domainErrorsClassName, importModuleSpecifier: '../../Domain' }
    ];

    const repos: string[] = [];

    for (const r of feature.domain.aggregateRoots) {
      repos.push(r.infraRepositoryName);
      imports.push(
        { name: r.name, importModuleSpecifier: `../../Domain/${r.name}/${r.name}` },
        { name: r.infraRepositoryName, importModuleSpecifier: `./${r.name}/${r.infraRepositoryName}` }
      );

      for (const e of r.entities) {
        repos.push(e.infraRepositoryName);
        imports.push({ name: e.infraRepositoryName, importModuleSpecifier: `./${r.name}/${e.infraRepositoryName}` });
      }
    }

    return this.moduleClassTransformer.transform({
      imports,
      extraStatementProvider(importedIdentifierMapper) {
        const classIdentifier = importedIdentifierMapper("DomainInfraModuleHelper");
        const methodIdentifier = ts.factory.createIdentifier("createMeta");

        const optionsObject = ts.factory.createObjectLiteralExpression([
          ts.factory.createPropertyAssignment("featureName", ts.factory.createStringLiteral(feature.name)),
          ts.factory.createPropertyAssignment("aggregateRootCtrs", ts.factory.createArrayLiteralExpression(feature.domain.aggregateRoots.map((r) => ts.factory.createIdentifier(r.name)))),
          ts.factory.createPropertyAssignment("domainErrors", ts.factory.createIdentifier(domainErrorsClassName)),
        ]);

        const createMeta = TsTransfromerHelper.createConstStatement("HcDomainInfraModuleMetaExtra", ts.factory.createCallExpression(
          ts.factory.createPropertyAccessExpression(classIdentifier, methodIdentifier),
          undefined,
          [optionsObject]
        ));

        return [
          TsTransfromerHelper.createConstStatement("HcDomainInfraAggrgateRootRepositories",
            ts.factory.createArrayLiteralExpression(repos.map((r) => ts.factory.createIdentifier(r)))
          ),
          createMeta
        ];
      },

      extraMetaProvider() {
        return ts.factory.createIdentifier("HcDomainInfraModuleMetaExtra");
      },
      source,
      context: context.tsContext,
    });
  }

  public supports(sourcefilePath: string, feature: FeatureMeta): boolean {
    return sourcefilePath.endsWith("DomainInfraModule.ts");
  }
}