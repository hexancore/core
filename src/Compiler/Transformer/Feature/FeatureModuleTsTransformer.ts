import ts from 'typescript';
import type {FeatureMeta } from "../../../Util/Feature/Meta/FeatureMeta";
import { AbstractFeatureTsTransformer } from "./AbstractFeatureTsTransformer";
import type { ProviderModuleMetaTransformDef } from '../ModuleClassTsTransformer';
import type { FeatureTransformContext } from './FeatureTransformContext';
import type { FeatureApplicationMessageMeta } from '../../../Util/Feature/Meta';
import type { FeatureSourcePath } from "../../../Util/Feature/FeatureModuleDiscoverer";

/**
 * Adding automatic injection of message handlers, services, infra module to `[Feature]Module` source.
 * Less write, more fun !
 */
export class FeatureModuleTsTransformer extends AbstractFeatureTsTransformer {

  public supports(featureSourcePath: FeatureSourcePath, feature: FeatureMeta): boolean {
    return featureSourcePath.localSourcePath.endsWith(feature.name + "Module.ts");
  }

  public transform(source: ts.SourceFile, context: FeatureTransformContext): ts.SourceFile {
    const feature = context.feature;

    const messageHandlersProviders: ProviderModuleMetaTransformDef[] = [];
    messageHandlersProviders.push(...this.createMessageHandlerProviders(feature.application.commands));
    messageHandlersProviders.push(...this.createMessageHandlerProviders(feature.application.queries));

    return this.moduleClassTransformer.transform({
      imports: [],
      meta: {
        imports: [],
        providers: [...messageHandlersProviders, ...this.createServiceProviders(feature)],
      },
      source,
      context: context.tsContext,
    });
  }

  private createMessageHandlerProviders(messages: FeatureApplicationMessageMeta[]): ProviderModuleMetaTransformDef[] {
    const providers: ProviderModuleMetaTransformDef[] = [];
    for (const m of messages) {
      const importPath = `./${m.path}/${m.handlerClass}`;
      providers.push({
        addToExports: false,
        name: m.handlerClass,
        importFrom: importPath
      });
    }

    return providers;
  }

  private createServiceProviders(feature: FeatureMeta): ProviderModuleMetaTransformDef[] {
    const providers: ProviderModuleMetaTransformDef[] = [];
    for (const s of feature.application.services) {
      if (!s.isInjectable) {
        continue;
      }

      providers.push({
        addToExports: false,
        name: s.name,
        importFrom: `./${s.path}`
      });
    }

    return providers;
  }
}