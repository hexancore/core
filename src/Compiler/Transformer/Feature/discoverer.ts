import { FeatureModuleDiscoverer } from "../../../Util/Feature/FeatureModuleDiscoverer";
// node ./lib/Compiler/Transformer/Feature/discoverer.js ./test/helper/libs/test-lib/src
async function run() {
  if (process.argv.length < 3) {
    throw new Error("Wrong execute, not given sourceRoot argument");
  }
  const sourceRoot = process.argv[2];
  const discoverer = new FeatureModuleDiscoverer(sourceRoot);
  const features = await discoverer.discoverAll();
  console.log(JSON.stringify(Array.from(features.entries())));
}

run();