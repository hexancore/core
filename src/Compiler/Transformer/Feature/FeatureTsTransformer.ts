import { FsHelper } from "@/Util/Filesystem/FsHelper";
import { execFileSync } from "node:child_process";
import ts from 'typescript';
import { FeatureMap, FeatureMeta } from "../../../Util/Feature/Meta/FeatureMeta";
import { TsTransfromerHelper } from "../TsTransformerHelper";
import type { AbstractFeatureTsTransformer } from "./AbstractFeatureTsTransformer";
import { FeatureInfraDomainModuleTsTransformer } from "./FeatureInfraDomainTsTransformer";
import { FeatureModuleTsTransformer } from "./FeatureModuleTsTransformer";
import path from "node:path";
import { FeatureModuleDiscoverer } from "../../../Util/Feature/FeatureModuleDiscoverer";
import { TsImportHelper } from "../TsImportHelper";
import { ModuleClassTsTransformer } from "../ModuleClassTsTransformer";
import type { FeatureTransformContext } from "./FeatureTransformContext";


/**
 * Adding automatic injection of message handlers, services, infra module to `[Feature]Module` source.
 * Less write, more fun !
 */
export class FeatureTsTransformer {
  private transformers: AbstractFeatureTsTransformer[];

  public constructor(
    private sourceRoot: string,
    private features: FeatureMap,
    importHelper: TsImportHelper,
  ) {

    const helpers = {
      importHelper: importHelper,
      moduleClassTransformer: new ModuleClassTsTransformer(importHelper),
      features: this.features,
    };

    this.transformers = [
      new FeatureInfraDomainModuleTsTransformer(helpers),
      new FeatureModuleTsTransformer(helpers),
    ];
  }

  public static create(sourceRoot: string, compilerDir?: string, features?: FeatureMap, needFixImportAccess = true): FeatureTsTransformer {
    sourceRoot = FsHelper.normalizePathSep(sourceRoot);
    compilerDir = compilerDir ?? FsHelper.normalizePathSep(path.dirname(path.dirname(__dirname)));
    features = features ?? this.loadFeaturesMapSync(sourceRoot, compilerDir);

    const importFromMapper = TsTransfromerHelper.createImportFromMapper(sourceRoot);
    const importHelper = new TsImportHelper(importFromMapper, needFixImportAccess);

    return new this(sourceRoot, features, importHelper);
  }

  private static loadFeaturesMapSync(sourceRoot: string, compilerDir: string): Map<string, FeatureMeta> {
    // TS transformer can't use async, but we need async code and execSync allows run async code :) Hack of day !
    const output = execFileSync('node', [`${compilerDir}/Transformer/Feature/discoverer.js`, sourceRoot], {
      windowsHide: true,
      shell: false,
      timeout: 60 * 1000
    });
    const featuresPlain = JSON.parse(output.toString());
    return FeatureMeta.parseMap(featuresPlain);
  }

  public transform(source: ts.SourceFile, context: ts.TransformationContext): ts.SourceFile {
    const sourceFilePath = FsHelper.normalizePathSep(source.fileName);
    const feature = this.getFeatureOfPath(sourceFilePath);
    if (!feature) {
      return source;
    }

    const sourceFilePathWithoutRoot = sourceFilePath.substring(this.sourceRoot.length + 1);
    const transformContext: FeatureTransformContext = {
      feature,
      sourceFilePathWithoutRoot,
      tsContext: context,
      diagnostics: [],
    };
    for (const t of this.transformers) {
      if (t.supports(sourceFilePathWithoutRoot, feature)) {
        const transformed = t.transform(source, transformContext);
        return transformed;
      }
    }

    return source;
  }

  private getFeatureOfPath(sourceFilePath: string): FeatureMeta | null {
    const featureName = FeatureModuleDiscoverer.extractFeatureNameFromPath(this.sourceRoot, sourceFilePath);
    if (!featureName) {
      return null;
    }

    return this.features.get(featureName) as FeatureMeta;
  }

  public supports(sourceFilePath: string, featureName: string): boolean {
    const feature = this.features.get(featureName)!;
    for (const t of this.transformers) {
      if (t.supports(sourceFilePath, feature)) {
        return true;
      }
    }

    return false;
  }
}