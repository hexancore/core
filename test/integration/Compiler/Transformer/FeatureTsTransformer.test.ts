/**
 * @group integration
 */

import { TsTransformerTestHelper } from "@/Compiler/Test/TsTransformerTestHelper";
import { FeatureModuleTsTransformer } from "@/Compiler/Transformer/Feature/FeatureModuleTsTransformer";
import { FeatureTsTransformer } from "@/Compiler/Transformer/Feature/FeatureTsTransformer";
import { readFileSync } from "fs";
import ts from "typescript";

describe(FeatureModuleTsTransformer.constructor.name, () => {
  const helper = TsTransformerTestHelper.createFromTsConfig(process.cwd() + "/tsconfig.json");
  let transformer: FeatureTsTransformer;

  beforeAll(() => {
    const sourceRoot = process.cwd() + "/test/helper/libs/test-lib/src";
    const compilerRoot = process.cwd() + "/lib/Compiler";
    transformer = FeatureTsTransformer.create(sourceRoot, compilerRoot);
  });

  test("transform FeatureModule", () => {
    const codeFilePath = process.cwd() + "/test/helper/libs/test-lib/src/Book/BookModule.ts";
    const sourceFile = helper.createSourceFileFromExisting(codeFilePath);
    const tsConfig = JSON.parse(readFileSync(process.cwd() + "/tsconfig.json").toString());

    const out = ts.transpileModule(sourceFile.text, {
      compilerOptions: tsConfig.compilerOptions,
      fileName: codeFilePath,
      transformers: {
        before: [(context) => (sourceFile: ts.SourceFile) => transformer.transform(sourceFile, context)]
      }
    });

    expect(out.outputText).toMatchSnapshot();
  });

  test.skip("transform HObject: Command", () => {
    const sourceFilePath = process.cwd() + "/test/helper/libs/test-lib/src/Book/Application/Book/Command/Create/BookCreateCommand.ts";

    const out = helper.transpileModule(transformer.transform.bind(transformer), sourceFilePath);

    expect(out.outputText).toMatchSnapshot();
  });

});


