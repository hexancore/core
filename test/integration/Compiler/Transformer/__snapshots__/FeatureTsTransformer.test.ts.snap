// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Function transform FeatureModule 1`] = `
""use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BookModule = exports.SOME_ASSIGMENT_NOISE = void 0;
const tslib_1 = require("tslib");
const common_1 = require("@nestjs/common");
const BookInfraModule_1 = require("./Infrastructure/BookInfraModule");
const BookCreateCommandHandler_1 = require("./Application/Book/Command/Create/BookCreateCommandHandler");
const BookGetByIdQueryHandler_1 = require("./Application/Book/Query/GetById/BookGetByIdQueryHandler");
const BookService_1 = require("./Application/Book/Service/BookService");
// TEST TRANSFORMING IGNORE COMMENT
exports.SOME_ASSIGMENT_NOISE = 10;
let BookModule = class BookModule {
};
exports.BookModule = BookModule;
exports.BookModule = BookModule = tslib_1.__decorate([
    (0, common_1.Module)({
        imports: [BookInfraModule_1.BookInfraModule],
        providers: [BookCreateCommandHandler_1.BookCreateCommandHandler, BookGetByIdQueryHandler_1.BookGetByIdQueryHandler, BookService_1.BookService],
        exports: []
    })
], BookModule);
//# sourceMappingURL=BookModule.js.map"
`;

exports[`Function transform HObject: Command 1`] = `
""use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BookCreateCommand = void 0;
class BookCreateCommand {
    constructor(title) {
        this.title = title;
    }
}
exports.BookCreateCommand = BookCreateCommand;
//# sourceMappingURL=BookCreateCommand.js.map"
`;
